ProcessLib/ComponentTransport/ComponentTransportFEM.h

```diff
@@ -341,35 +344,41 @@ class LocalAssemblerData : public ComponentTransportLocalAssemblerInterface
                            solute_dispersivity_transverse *
                                velocity_magnitude) *
                           I);
+            const double R_times_phi(retardation_factor * porosity);
+            GlobalDimVectorType const mass_density_flow = velocity * density;
+            auto const N_t_N = (N.transpose() * N).eval();
+            if (_process_data.non_advective_form)
+            {
+                const double C_int_pt(N.dot(C_nodal_values));
+                MCp.noalias() += N_t_N * (C_int_pt * R_times_phi * drho_dp * w);
+                MCC.noalias() += N_t_N * (C_int_pt * R_times_phi * drho_dC * w);
+                KCC.noalias() -= dNdx.transpose() * mass_density_flow * N * w;
+            }
+            else
+            {
+                BC.noalias() -=
+                    N.transpose() *
+                    (mass_density_flow.dot(dNdx * C_nodal_values) * w);
+            }
+            MCC.noalias() += N_t_N * (R_times_phi * density * w);
+            KCC.noalias() += dNdx.transpose() * hydrodynamic_dispersion * dNdx *
+                                 (density * w) +
+                             N_t_N * (decay_rate * porosity *
+                                      retardation_factor * density * w);
 
-            // matrix assembly
-            MCp.noalias() += w * N.transpose() * N * C_nodal_values *
-                             retardation_factor * porosity * drho_dp * N;
-            MCC.noalias() += w * N.transpose() * retardation_factor * porosity *
-                                 density * N +
-                             w * N.transpose() * retardation_factor * porosity *
-                                 N * C_nodal_values * drho_dC * N;
-
-            KCC.noalias() +=
-                (-dNdx.transpose() * velocity * density * N +
-                 dNdx.transpose() * density * hydrodynamic_dispersion * dNdx +
-                 N.transpose() * decay_rate * porosity * retardation_factor *
-                     density * N) *
-                w;
-
-            MpC.noalias() += w * N.transpose() * porosity * drho_dC * N;
+            MpC.noalias() += N_t_N * (porosity * drho_dC * w);
 
             // Calculate Mpp, Kpp, and bp in the first loop over components
             if (component_id == 0)
             {
-                Mpp.noalias() += w * N.transpose() * porosity * drho_dp * N;
+                Mpp.noalias() += N_t_N * (porosity * drho_dp * w);
                 Kpp.noalias() +=
-                    w * dNdx.transpose() * density * K_over_mu * dNdx;
+                    dNdx.transpose() * (density * K_over_mu * w) * dNdx;
 
                 if (_process_data.has_gravity)
                 {
-                    Bp += w * density * density * dNdx.transpose() * K_over_mu *
-                          b;
+                    Bp.noalias() += dNdx.transpose() *
+                                    (density * density * K_over_mu * b * w);
```

Comment:
If you want to optimize for performance you could move the `density * density` to `w` (`K_over_mu * b` is a vector).

Reply:
thanks - moved K_over_mu * b